src/main/java/jalf/compiler/Cog.java

```diff
@@ -47,147 +48,157 @@
     protected Relation expr;
 
     public Cog(Relation expr) {
-        super();
-        this.expr = expr;
+	super();
+	this.expr = expr;
     }
 
     public Relation getExpr() {
-        return expr;
+	return expr;
     }
 
     public abstract Stream<Tuple> stream();
 
     public Cog select(Select relation) {
-        Selection selection = relation.getSelection();
+	Selection selection = relation.getSelection();
 
-        Supplier<Stream<Tuple>> supplier = () -> this.stream()
-                .map(t -> selection.apply(t))
-                .distinct();
+	Supplier<Stream<Tuple>> supplier = () -> this.stream()
+		.map(t -> selection.apply(t)).distinct();
 
-        return new BaseCog(relation, supplier);
+	return new BaseCog(relation, supplier);
     }
 
     /** Default compilation of `project`. */
     public Cog project(Project projection) {
-        AttrList on = projection.getAttributes();
-        TupleType tt = projection.getTupleType();
-
-        // stream compilation: map projection + distinct
-        Supplier<Stream<Tuple>> supplier = () -> this.stream()
-                .map(t -> t.project(on, tt))
-                .distinct();
-
-        return new BaseCog(projection, supplier);
+	AttrList on = projection.getAttributes();
+	TupleType tt = projection.getTupleType();
+	Keys keys = projection.getOperand().getKeys();
+	// Key key = keys.toList().get(0);
+	Supplier<Stream<Tuple>> supplier;
+	// avoid duplicate
+	if (keys.stream().anyMatch((Key k) -> k.intersect(on).equals(k))) {
```

Comment:
Wrong: `k.intersect(on).equals(k)` is always false. k.intersect returns an AttrList while k is a Key => TypeError